spec.bs

```diff
@@ -232,7 +232,161 @@ to conduct an auction to select an advertisement to display to the user, they ca
+specification are those that explicitly list the global names provided here.
+
+<pre class="idl">
+[Exposed=InterestGroupBiddingScriptRunner,
+Global=(InterestGroupScriptRunner, InterestGroupBiddingScriptRunner)]
+interface InterestGroupBiddingScriptRunnerGlobalScope{
+  boolean setBid();
+  boolean setBid(GenerateBidOutput generateBidOutput);
+  undefined setPriority(double priority);
+  undefined setPrioritySignalsOverride(DOMString key, double priority);
+};
+
+[Exposed=InterestGroupScoringScriptRunner,
+Global=(InterestGroupScriptRunner, InterestGroupScoringScriptRunner)]
+interface InterestGroupScoringScriptRunnerGlobalScope{
+};
+
+[Exposed=InterestGroupReportingScriptRunner,
+Global=(InterestGroupScriptRunner, InterestGroupReportingScriptRunner)]
+interface InterestGroupReportingScriptRunnerGlobalScope{
+  undefined sendReportTo(DOMString url);
+};
+
+dictionary AdSize {
+  required double width;
+  required DOMString widthUnits;
+  required double height;
+  required DOMString heightUnits;
+};
+
+dictionary AdDescriptor {
+  required DOMString url;
+  AdSize adSize;
+};
+
+dictionary GenerateBidOutput {
+  required double bid;
+  required (DOMString or AdDescriptor) adDescriptor;
+  DOMString ad;
+  sequence<(DOMString or AdDescriptor)> adComponentDescriptors;
+  double adCost;
+  double modelingSignals;
+  boolean allowComponentAuction = false;
+};
+</pre>
+
+Issue: TODO: Turn the follwing methods into spec language.
+
+The setBid() method steps are:
+1. Clear the bid.
+1. Return true.
+
+The setBid(|generateBidOutput|) method steps are (TODO: Finish this. Some steps are not in spec
+language yet):
+1. Clear the bid.
+1. If |generateBidOutput| is null or undefined, return true.
+1. If |generateBidOutput|["{{GenerateBidOutput/bid}}"] is less than or equal to 0, return true.
+1. If it's a component auction, but
+  |generateBidOutput|["{{GenerateBidOutput/allowComponentAuction}}"] is false, return false.
+1. Let |bidToSet| be a new [=generated bid=].
+1. Set |bidToSet|'s [=generated bid/bid=] to |generateBidOutput|["{{GenerateBidOutput/bid}}"].
+1. If |generateBidOutput|["{{GenerateBidOutput/ad}}"] [=map/exists=]:
+  1. Let |adJson| be the result of [=serializing a JavaScript value to a JSON string=], given
+    |generateBidOutput|["{{GenerateBidOutput/ad}}"].
+  1. If |adJson| is failure, return failure.
+  1. Set |bidToSet|'s [=generated bid/ad=] to |adJson|.
+1. Let |renderURL| be null.
+1. If |generateBidOutput|["{{GenerateBidOutput/adDescriptor}}"] is a {{DOMString}}:
+  1. Let |renderURL| be the result of running the [=URL parser=] on
+    |generateBidOutput|["{{GenerateBidOutput/adDescriptor}}"].
+1. Otherwise:
+  1. TODO: Let |adDescriptor| be the result of [=converting an ad descriptor=] given
+    |generateBidOutput|["{{GenerateBidOutput/adDescriptor}}"].
+  1. If |adDescriptor| is failure, return false.
+  1. Let |parsedAdDescriptor| be the result of parsing |adDescriptor|.

```

Comment:
What does it mean to "parse" something coming from [=converting an ad descript=]? is this already accomplished by the [=converting an ad descriptor=] call above?

Reply:
Uploaded a new commit this morning that fixed many problems in the old commit about setBid()'s steps, and most comments are obsolete now. I think the new commit is in a good shape for review. Sorry about having you looking at a version that had quite a few problems.